File: python/18_error_handling.py
# ============================================================
# Topic:  Error Handling  (FOCUS: try/except, raise, finally, propagate)
# Summary: Catch exceptions, raise your own, clean up with finally.
# Run:     python python/18_error_handling.py
#
# KEY DIFFERENCES:
#   Python:  try/except/else/finally + raise (stdlib)
#   TS/JS:   try/catch/finally + throw
#   Rust:    Result<T,E> + ? operator | panic!
#   Java:    try/catch/finally + throw
#   C++:     try/catch + throw
# ============================================================


# ----- 1. try/except — catch and recover -----
def safe_divide(a, b):
    try:
        return a / b
    except ZeroDivisionError:
        print("caught: division by zero", flush=True)  # Output: caught: division by zero
        return None


# ----- 2. raise — signal an error -----
def validate_positive(n):
    if n < 0:
        raise ValueError("must be positive")
    return n


# ----- 3. Custom Exception class -----
class CustomBad(Exception):
    """Raised when input fails a custom rule."""


def check_valid(label):
    if label == "invalid":
        raise CustomBad("invalid")


# ----- 4. finally — always runs -----
def with_finally():
    try:
        raise RuntimeError("trigger")
    except RuntimeError:
        pass
    finally:
        print("finally ran", flush=True)  # Output: finally ran


# ----- 5. Re-raise / propagate — outer handler catches inner -----
def inner_fail():
    raise ValueError("inner failed")


def outer_catch():
    try:
        inner_fail()
    except ValueError as e:
        print(f"outer: {e}", flush=True)  # Output: outer: inner failed


# ----- 6. else clause — runs only when try succeeds -----
def parse_int(text):
    try:
        value = int(text)
    except ValueError:
        return None
    else:
        return value * 2  # no exception → else block runs


def main():
    # ----- 1. try/except -----
    result = safe_divide(20, 2)
    print(f"result: {int(result)}", flush=True)  # Output: result: 10
    safe_divide(10, 0)

    # ----- 2. raise (used by validate_positive; no extra output) -----
    validate_positive(5)

    # ----- 3. Custom Exception -----
    try:
        check_valid("invalid")
    except CustomBad as e:
        print(f"CustomBad: {e}", flush=True)  # Output: CustomBad: invalid

    # ----- 4. finally -----
    with_finally()

    # ----- 5. Re-raise / propagate -----
    outer_catch()

    # ----- 6. else clause (demonstrated; no extra output) -----
    parse_int("7")


if __name__ == "__main__":
    main()

# Cross-language note:
#   Python's bare raise re-raises the current exception; use `raise X from Y` to chain causes.
#   else runs only when try completes without exception; finally always runs.
#   Custom exceptions subclass Exception — like Java checked/unchecked or TS custom Error classes.
#   Rust prefers Result over exceptions; C++ and Java use similar try/catch patterns.

File: rust/18_error_handling.rs
// ============================================================
// Topic:  Error Handling
// Summary: Result<T,E>, ? operator, panic!/catch_unwind, custom error enum, match on Err.
// Run:     rustc --edition 2021 rust/18_error_handling.rs -o rust/18_error_handling.exe && rust/18_error_handling.exe
//
// KEY DIFFERENCES:
//   Python:  try/except, raise, finally
//   TS/JS:   try/catch/finally, throw
//   Rust:    Result<T,E> (no exceptions); panic! for unrecoverable; ? propagates Err
//   Java:    try/catch/finally, throws, checked exceptions
//   C++:     try/catch, throw, std::optional / std::expected (C++23)
// ============================================================

// ----- 1. Result<T,E> — Ok/Err instead of throw/catch -----
fn divide(a: i32, b: i32) -> Result<i32, String> {
    if b == 0 {
        Err("division by zero".to_string()) // Python: raise ValueError("division by zero")
    } else {
        Ok(a / b)
    }
}

// ----- 2. Helper prints "caught:" on Err (Python: except block) -----
fn run_with_catch<F>(f: F)
where
    F: FnOnce() -> Result<i32, String>,
{
    match f() {
        Ok(v) => println!("result: {}", v),
        Err(e) => println!("caught: {}", e), // Output: caught: division by zero
    }
}

// ----- 3. finally pattern via Drop (runs when guard goes out of scope) -----
struct Finally;
impl Drop for Finally {
    fn drop(&mut self) {
        println!("finally ran"); // Output: finally ran   | Python: finally block
    }
}

// ----- 4. Custom error enum — match on Err variants -----
#[derive(Debug)]
enum AppError {
    CustomBad(String),
}

fn validate(input: &str) -> Result<(), AppError> {
    if input.is_empty() {
        Err(AppError::CustomBad("invalid".to_string()))
    } else {
        Ok(())
    }
}

// ----- 5. ? operator — propagates Err up the call stack -----
fn inner() -> Result<i32, String> {
    let _ = divide(5, 0)?; // ? returns early with Err if divide fails
    Ok(10)
}

fn outer() -> Result<(), String> {
    match inner() {
        Ok(_) => Ok(()),
        Err(_) => {
            println!("outer: inner failed"); // Output: outer: inner failed
            Err("inner failed".to_string())
        }
    }
}

fn main() {
    // ----- 1. match on Ok — successful Result -----
    match divide(20, 2) {
        Ok(v) => println!("result: {}", v), // Output: result: 10
        Err(e) => println!("caught: {}", e),
    }

    // ----- 2. Helper prints "caught:" on Err -----
    run_with_catch(|| divide(10, 0)); // Output: caught: division by zero

    // ----- 3. Custom error enum — match on Err variant -----
    match validate("") {
        Ok(()) => {}
        Err(AppError::CustomBad(msg)) => println!("CustomBad: {}", msg), // Output: CustomBad: invalid
    }

    // ----- 4. finally via Drop — scoped guard runs cleanup on exit -----
    {
        let _guard = Finally;
    } // Output: finally ran

    // ----- 5. ? propagates from inner; outer handles the failure -----
    let _ = outer(); // Output: outer: inner failed

    // ----- 6. panic!/catch_unwind (optional) — isolate panics without crashing stdout -----
    // panic::catch_unwind(|| panic!("demo")); // catches unwinding panics; no direct Python equivalent
    // Cross-language: Rust prefers Result for recoverable errors; panic! for bugs/invariants.
}